File: backend/app/services/rules_engine.py
"""Rule Engine JSON generation (Sprint 7).

Turns extracted scenarios into structured :class:`Rule` objects. Each scenario
becomes one rule: the condition clause is parsed into typed facts/operators,
and the resolution group maps to a canonical action.

``to_engine_json`` renders a :class:`RuleSet` into the portable JSON shape
consumed by common rule engines (a ``conditions.all`` array + an ``event``).
"""

from __future__ import annotations

import re

from app.models.schemas import (
    PipelineResult,
    Rule,
    RuleAction,
    RuleCondition,
    RuleOperator,
    RuleSet,
    Scenario,
)
from app.services.storage import new_id

# Comparison phrases → operators. Order matters (longest/﻿specific first).
_COMPARATORS: list[tuple[re.Pattern[str], RuleOperator]] = [
    (re.compile(r"\b(?:greater than or equal|at least|no less than)\b"), RuleOperator.greater_than),
    (re.compile(r"\b(?:less than or equal|at most|no more than|within)\b"), RuleOperator.less_than),
    (re.compile(r"\b(?:greater than|more than|over|above|exceeds?)\b"), RuleOperator.greater_than),
    (re.compile(r"\b(?:less than|fewer than|under|below)\b"), RuleOperator.less_than),
    (re.compile(r"\b(?:is not|are not|isn't|aren't|not equal)\b"), RuleOperator.not_equals),
    (re.compile(r"\b(?:contains?|includes?|has)\b"), RuleOperator.contains),
    (re.compile(r"\b(?:is|are|equals?|equal to)\b"), RuleOperator.equals),
]

_NUMBER = re.compile(r"(-?\d+(?:\.\d+)?)")
_FACT_STOPWORDS = {"the", "a", "an", "is", "are", "of", "and", "or", "customer"}

_GROUP_ACTIONS: dict[str, RuleAction] = {
    "Auto-approve": RuleAction(type="set", target="status", value="approved"),
    "Deny": RuleAction(type="set", target="status", value="denied"),
    "Escalate": RuleAction(type="route", target="queue", value="manual_review"),
    "Notify": RuleAction(type="notify", target="stakeholder"),
    "Record": RuleAction(type="log", target="audit_trail"),
}


def _slugify_fact(text: str) -> str:
    words = [
        w
        for w in re.findall(r"[a-zA-Z]+", text.lower())
        if w not in _FACT_STOPWORDS
    ]
    return "_".join(words[:4]) or "condition"


def _parse_condition(condition: str) -> RuleCondition:
    """Best-effort parse of a natural-language condition into a typed fact."""
    operator = RuleOperator.equals
    for pattern, op in _COMPARATORS:
        if pattern.search(condition):
            operator = op
            break

    number = _NUMBER.search(condition)
    if number and operator in {
        RuleOperator.greater_than,
        RuleOperator.less_than,
    }:
        value: object = float(number.group(1))
        if float(value).is_integer():
            value = int(value)
        fact = _slugify_fact(_NUMBER.sub("", condition))
    else:
        # Use trailing noun-ish token as the value for equality checks.
        tokens = re.findall(r"[a-zA-Z]+", condition)
        value = tokens[-1] if tokens else "true"
        fact = _slugify_fact(condition)

    return RuleCondition(fact=fact, operator=operator, value=value)


def scenario_to_rule(scenario: Scenario, priority: int) -> Rule:
    condition = _parse_condition(scenario.condition)
    action = _GROUP_ACTIONS.get(
        scenario.resolution_group,
        RuleAction(type="set", target="outcome", value="review"),
    )
    return Rule(
        id=new_id("rul"),
        name=scenario.resolution[:80] or "Generated rule",
        description=f"{scenario.condition} → {scenario.resolution}",
        priority=priority,
        all=[condition],
        actions=[action],
        enabled=scenario.confidence >= 0.6,
        source_scenario_id=scenario.id,
    )


def generate_rule_set(project_id: str, results: list[PipelineResult]) -> RuleSet:
    rules: list[Rule] = []
    priority = 10
    for result in results:
        for scenario in result.scenarios:
            rules.append(scenario_to_rule(scenario, priority))
            priority += 10

    return RuleSet(
        id=new_id("rst"),
        project_id=project_id,
        name="Generated rule set",
        version="1.0.0",
        rules=rules,
    )


def to_engine_json(rule_set: RuleSet) -> dict:
    """Render a rule set into portable Rule Engine JSON."""
    return {
        "name": rule_set.name,
        "version": rule_set.version,
        "rules": [
            {
                "id": rule.id,
                "name": rule.name,
                "priority": rule.priority,
                "conditions": {
                    "all": [
                        {
                            "fact": c.fact,
                            "operator": c.operator.value,
                            "value": c.value,
                        }
                        for c in rule.all
                    ]
                },
                "event": {
                    "type": "ruleforge.match",
                    "params": {
                        "actions": [a.model_dump(exclude_none=True) for a in rule.actions]
                    },
                },
            }
            for rule in rule_set.rules
            if rule.enabled
        ],
    }

File: src/server/rules-engine.ts
/**
 * Rule Engine JSON generation (Sprint 7) — ported from `services/rules_engine.py`.
 *
 * Turns extracted scenarios into structured Rule objects: the condition clause
 * is parsed into a typed fact/operator/value, and the resolution group maps to
 * a canonical action. `toEngineJson` renders a RuleSet into portable Rule
 * Engine JSON (a `conditions.all` array + an `event`).
 */

import type {
  PipelineResult,
  Rule,
  RuleAction,
  RuleCondition,
  RuleOperator,
  RuleSet,
  Scenario,
} from "@/lib/types";
import { newId, nowIso } from "@/server/id";

const COMPARATORS: [RegExp, RuleOperator][] = [
  [/\b(?:greater than or equal|at least|no less than)\b/i, "greater_than"],
  [/\b(?:less than or equal|at most|no more than|within)\b/i, "less_than"],
  [/\b(?:greater than|more than|older than|longer than|over|above|exceeds?)\b/i, "greater_than"],
  [/\b(?:less than|fewer than|younger than|newer than|sooner than|under|below)\b/i, "less_than"],
  [/\b(?:is not|are not|isn't|aren't|not equal)\b/i, "not_equals"],
  [/\b(?:contains?|includes?|has)\b/i, "contains"],
  [/\b(?:is|are|equals?|equal to)\b/i, "equals"],
];

const NUMBER = /(-?\d+(?:\.\d+)?)/;
const FACT_STOPWORDS = new Set([
  "the",
  "a",
  "an",
  "is",
  "are",
  "of",
  "and",
  "or",
  "customer",
]);

const GROUP_ACTIONS: Record<string, RuleAction> = {
  "Auto-approve": { type: "set", target: "status", value: "approved" },
  Deny: { type: "set", target: "status", value: "denied" },
  Escalate: { type: "route", target: "queue", value: "manual_review" },
  Notify: { type: "notify", target: "stakeholder" },
  Record: { type: "log", target: "audit_trail" },
};

function slugifyFact(text: string): string {
  const words = (text.toLowerCase().match(/[a-z]+/g) ?? []).filter(
    (w) => !FACT_STOPWORDS.has(w)
  );
  return words.slice(0, 4).join("_") || "condition";
}

function parseCondition(condition: string): RuleCondition {
  let operator: RuleOperator = "equals";
  for (const [pattern, op] of COMPARATORS) {
    if (pattern.test(condition)) {
      operator = op;
      break;
    }
  }

  const number = NUMBER.exec(condition);
  if (number && (operator === "greater_than" || operator === "less_than")) {
    const parsed = Number(number[1]);
    const value = Number.isInteger(parsed) ? parsed : parsed;
    const fact = slugifyFact(condition.replace(NUMBER, ""));
    return { fact, operator, value };
  }

  const tokens = condition.match(/[a-zA-Z]+/g) ?? [];
  const value = tokens.length > 0 ? tokens[tokens.length - 1] : "true";
  return { fact: slugifyFact(condition), operator, value };
}

export function scenarioToRule(scenario: Scenario, priority: number): Rule {
  const condition = parseCondition(scenario.condition);
  const action = GROUP_ACTIONS[scenario.resolution_group] ?? {
    type: "set",
    target: "outcome",
    value: "review",
  };
  return {
    id: newId("rul"),
    name: scenario.resolution.slice(0, 80) || "Generated rule",
    description: `${scenario.condition} → ${scenario.resolution}`,
    priority,
    all: [condition],
    actions: [action],
    enabled: scenario.confidence >= 0.6,
    source_scenario_id: scenario.id,
  };
}

export function generateRuleSet(
  projectId: string,
  results: PipelineResult[]
): RuleSet {
  const rules: Rule[] = [];
  let priority = 10;
  for (const result of results) {
    for (const scenario of result.scenarios) {
      rules.push(scenarioToRule(scenario, priority));
      priority += 10;
    }
  }

  return {
    id: newId("rst"),
    project_id: projectId,
    name: "Generated rule set",
    version: "1.0.0",
    rules,
    created_at: nowIso(),
  };
}

export function toEngineJson(ruleSet: RuleSet): unknown {
  return {
    name: ruleSet.name,
    version: ruleSet.version,
    rules: ruleSet.rules
      .filter((rule) => rule.enabled)
      .map((rule) => ({
        id: rule.id,
        name: rule.name,
        priority: rule.priority,
        conditions: {
          all: rule.all.map((c) => ({
            fact: c.fact,
            operator: c.operator,
            value: c.value,
          })),
        },
        event: {
          type: "ruleforge.match",
          params: {
            actions: rule.actions.map((a) =>
              a.value === undefined
                ? { type: a.type, target: a.target }
                : { type: a.type, target: a.target, value: a.value }
            ),
          },
        },
      })),
  };
}
